sdk/python/kfp/dsl/_container_op.py

```diff
@@ -249,6 +170,132 @@ def add_env_variable(self, env_variable):
+      command: the command to run in the container.
+          If None, uses default CMD in defined in container.
+      arguments: the arguments of the command. The command can include "%s" and supply
+          a PipelineParam as the string replacement. For example, ('echo %s' % input_param).
+          At container run time the argument will be 'echo param_value'.
+    """
+    # name will be generated when attached to ContainerOps
+    super().__init__(name, image)
+    self.command = command
+    self.arguments = arguments
+
+    matches = []
+    for arg in (command or []) + (arguments or []):
+      match = re.findall(r'{{pipelineparam:op=([\w\s_-]*);name=([\w\s_-]+);value=(.*?)}}', str(arg))
+      matches += match
+
+    self.argument_inputs = [_pipeline_param.PipelineParam(x[1], x[0], x[2])
+                            for x in list(set(matches))]
+
+
+class ContainerOp(ContainerBase):
+  """Represents an op implemented by a docker container image."""
+
+  def __init__(self, name: str, image: str, command: str=None, arguments: str=None,
+               file_outputs : Dict[str, str]=None, is_exit_handler=False):
+    """Create a new instance of ContainerOp.
+
+    Args:
+      name: the name of the op. It does not have to be unique within a pipeline
+          because the pipeline will generates a unique new name in case of conflicts.
+      image: the container image name, such as 'python:3.5-jessie'
+      command: the command to run in the container.
+          If None, uses default CMD in defined in container.
+      arguments: the arguments of the command. The command can include "%s" and supply
+          a PipelineParam as the string replacement. For example, ('echo %s' % input_param).
+          At container run time the argument will be 'echo param_value'.
+      file_outputs: Maps output labels to local file paths. At pipeline run time,
+          the value of a PipelineParam is saved to its corresponding local file. It's
+          one way for outside world to receive outputs of the container.
+      is_exit_handler: Whether it is used as an exit handler.
+    """
+    if not _pipeline.Pipeline.get_default_pipeline():
+      raise ValueError('Default pipeline not defined.')
+
+    valid_name_regex = r'^[A-Za-z][A-Za-z0-9\s_-]*$'
+    if not re.match(valid_name_regex, name):
+      raise ValueError('Only letters, numbers, spaces, "_", and "-"  are allowed in name. Must begin with letter: %s' % (name))
+
+    # human_name must exist to construct containerOps name
+    self.human_name = name
+    super().__init__(_pipeline.Pipeline.get_default_pipeline().add_op(self, is_exit_handler), image)
+
+    self.command = command
+    self.arguments = arguments
+    self.is_exit_handler = is_exit_handler
+    self.node_selector = {}
+    self.volumes = []
+    self.pod_annotations = {}
+    self.pod_labels = {}
+    self.num_retries = 0
+    self.sidecars = []
+
+    matches = []
+    for arg in (command or []) + (arguments or []):
```

Comment:
We actually have a plan to extract the pipeline param from the whole body of the ContainerOp instead of just command and arguments. Argo supports placeholders for the whole template spec. If you don't want to expand the scope of the PR, you might create a utility method to extract param from a container's command and arguments and apply it to ContainerOp (for the main container) and sidecar containers here.

Reply:
@hongye-sun 
Hey, I have made a very big change to the PR to cater to this feature. 

Essentially, it is a recursive scan of `ContainerOp` as well as any k8s object and cached inside `inputs`.

I also replaced the `container` prop in argo workflow template with a wrapped k8s `V1Container`, so that any valid k8s prop can be set for `ContainerOp`.

Reply:
@eterna2, awesome, really appreciate your contribution here.